File: mobile/lib/core/models/user_profile.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// 사용자 프로필 모델
class UserProfile {
  final String uid;
  final String email;
  final String? displayName;
  final String? nickname;
  final String? photoURL;
  final String? provider; // 'email', 'google', 'apple'
  final int? handicap; // 핸디캡
  final String? defaultTee; // 기본 티셋
  final Map<String, bool>? termsAgreement; // 약관 동의 내역
  final DateTime createdAt;
  final DateTime updatedAt;

  UserProfile({
    required this.uid,
    required this.email,
    this.displayName,
    this.nickname,
    this.photoURL,
    this.provider,
    this.handicap,
    this.defaultTee,
    this.termsAgreement,
    required this.createdAt,
    required this.updatedAt,
  });

  Map<String, dynamic> toMap() {
    return {
      'uid': uid,
      'email': email,
      'displayName': displayName,
      'nickname': nickname,
      'photoURL': photoURL,
      'provider': provider,
      'handicap': handicap,
      'defaultTee': defaultTee,
      'termsAgreement': termsAgreement,
      'createdAt': Timestamp.fromDate(createdAt),
      'updatedAt': Timestamp.fromDate(updatedAt),
    };
  }

  factory UserProfile.fromMap(Map<String, dynamic> map) {
    return UserProfile(
      uid: map['uid'] as String,
      email: map['email'] as String,
      displayName: map['displayName'] as String?,
      nickname: map['nickname'] as String?,
      photoURL: map['photoURL'] as String?,
      provider: map['provider'] as String?,
      handicap: map['handicap'] as int?,
      defaultTee: map['defaultTee'] as String?,
      termsAgreement: (map['termsAgreement'] as Map<String, dynamic>?)
          ?.map((k, v) => MapEntry(k, v as bool)),
      createdAt: (map['createdAt'] as Timestamp).toDate(),
      updatedAt: (map['updatedAt'] as Timestamp).toDate(),
    );
  }

  UserProfile copyWith({
    String? uid,
    String? email,
    String? displayName,
    String? nickname,
    String? photoURL,
    String? provider,
    int? handicap,
    String? defaultTee,
    Map<String, bool>? termsAgreement,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return UserProfile(
      uid: uid ?? this.uid,
      email: email ?? this.email,
      displayName: displayName ?? this.displayName,
      nickname: nickname ?? this.nickname,
      photoURL: photoURL ?? this.photoURL,
      provider: provider ?? this.provider,
      handicap: handicap ?? this.handicap,
      defaultTee: defaultTee ?? this.defaultTee,
      termsAgreement: termsAgreement ?? this.termsAgreement,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: mobile-rn/src/core/types/userProfile.ts
import firestore, {
  type FirebaseFirestoreTypes,
} from '@react-native-firebase/firestore';

/**
 * 사용자 프로필 (Firestore users/{uid})
 * Flutter UserProfile와 동일 구조
 */
export interface UserProfile {
  uid: string;
  email: string;
  displayName: string | null;
  nickname: string | null;
  photoURL: string | null;
  provider: string | null; // 'email' | 'google' | 'apple'
  handicap?: number;
  defaultTee?: string | null;
  termsAgreement?: Record<string, boolean> | null;
  createdAt: Date;
  updatedAt: Date;
}

export interface TermsAgreement {
  serviceTerms: boolean;
  privacyPolicy: boolean;
  marketing: boolean;
  agreedAt: Date;
}

export function userProfileFromMap(
  map: Record<string, unknown>
): UserProfile {
  const toDate = (v: unknown): Date => {
    if (v && typeof (v as FirebaseFirestoreTypes.Timestamp).toDate === 'function') {
      return (v as FirebaseFirestoreTypes.Timestamp).toDate();
    }
    if (v instanceof Date) return v;
    return new Date();
  };
  return {
    uid: (map.uid as string) ?? '',
    email: (map.email as string) ?? '',
    displayName: (map.displayName as string | null) ?? null,
    nickname: (map.nickname as string | null) ?? null,
    photoURL: (map.photoURL as string | null) ?? null,
    provider: (map.provider as string | null) ?? null,
    handicap: map.handicap as number | undefined,
    defaultTee: (map.defaultTee as string | null) ?? null,
    termsAgreement: map.termsAgreement as Record<string, boolean> | null | undefined,
    createdAt: toDate(map.createdAt),
    updatedAt: toDate(map.updatedAt),
  };
}

export function userProfileToMap(profile: UserProfile): Record<string, unknown> {
  return {
    uid: profile.uid,
    email: profile.email,
    displayName: profile.displayName,
    nickname: profile.nickname,
    photoURL: profile.photoURL,
    provider: profile.provider,
    handicap: profile.handicap,
    defaultTee: profile.defaultTee,
    termsAgreement: profile.termsAgreement,
    createdAt: firestore.Timestamp.fromDate(profile.createdAt),
    updatedAt: firestore.Timestamp.fromDate(profile.updatedAt),
  } as Record<string, unknown>;
}
